httpserver/command_line.cpp

```diff
@@ -0,0 +1,128 @@
+    return options_;
+}
+
+void command_line_parser::parse_args(int argc, char* argv[]) {
+    std::string config_file;
+    po::options_description all, desc, hidden;
+    // clang-format off
+    desc.add_options()
+        ("lokid-key", po::value(&options_.lokid_key_path), "Path to the Service Node key file")
+        ("data-dir", po::value(&options_.data_dir), "Path to persistent data (defaults to ~/.loki/storage)")
+        ("config-file", po::value(&config_file), "Path to custom config file (defaults to `storage-server.conf' inside --data-dir)")
+        ("log-level", po::value(&options_.log_level), "Log verbosity level, see Log Levels below for accepted values")
+        ("lokid-rpc-port", po::value(&options_.lokid_rpc_port), "RPC port on which the local Loki daemon is listening")
+        ("force-start", po::bool_switch(&options_.force_start), "Ignore the initialisation ready check")
+        ("version,v", po::bool_switch(&options_.print_version), "Print the version of this binary")
+        ("help", "Shows this help message");
+        // Add hidden ip and port options.  You technically can use the `--ip=` and `--port=` with
+        // these here, but they are meant to be positional.  More usefully, you can specify `ip=`
+        // and `port=` in the config file to specify them.
+    hidden.add_options()
+        ("ip", po::value(&options_.ip), "IP to listen on")
+        ("port", po::value(&options_.port), "Port to listen on");
+    // clang-format on
+
+    all.add(desc).add(hidden);
+    po::positional_options_description pos_desc;
+    pos_desc.add("ip", 1);
+    pos_desc.add("port", 1);
+
+    std::string binary_name = fs::basename(argv[0]);
+
+    po::variables_map vm;
+    try {
+        po::store(po::command_line_parser(argc, argv)
+                      .options(all)
+                      .positional(pos_desc)
+                      .run(),
+                  vm);
+        po::notify(vm);
+    } catch (const boost::program_options::error& e) {
+        std::cerr << "Invalid options: " << e.what() << std::endl;
+        print_usage(desc, binary_name);
+        throw;
+    }
+
+    if (config_file.empty()) {
+        config_file =
+            (fs::path(options_.data_dir) / "storage-server.conf").string();
+    }
+    if (fs::exists(config_file)) {
+        try {
+            po::store(po::parse_config_file<char>(config_file.c_str(), all),
+                      vm);
+            po::notify(vm);
+        } catch (const boost::program_options::error& e) {
+            std::cerr << "Invalid options in config file: " << e.what()
+                      << std::endl;
+            print_usage(desc, binary_name);
+            throw;
+        }
+    }
+
+    if (!vm.count("data-dir")) {
+        if (auto home_dir = get_home_dir()) {
+            options_.data_dir = (home_dir.get() / ".loki" / "storage").string();
+        }
+    }
+
+    if (options_.print_version) {
```

Comment:
You already exit if `options.print_version`. No need to specify `early_exit_` here. I would personally remove `early_exit_` and instead check `options.help` in main to decide if you need to exit.

Reply:
I've changed the way the "early exit" code paths are handled in a new commit.
Basically the parser only populates the `command_line_options` structure, and leave it to main to take action (e.g. print the version or the usage).